doc/api18970.md

```diff
@@ -0,0 +1,175 @@
+
+## Background
+
+See also the Zephyr [glossary][].
+
+[glossary]: https://docs.zephyrproject.org/latest/glossary.html
+
+### Thread Terminology
+
+Zephyr defines [six thread
+states](https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-states)
+of which four are active:
+
+* **Ready** when there is nothing that prevents the thread from becoming
+  active as the current thread, but it is not the current thread.
+* **Running** when the thread is active as the current thread (on a
+  processor).
+* **Waiting** when the thread is on a queue waiting for an event to occur
+  that will transition it to *Ready*.
+* **Suspended** when the thread is inactive and must be transitioned to
+  *Ready* through explicitly (via `k_thread_resume()`.
+
+We shall say that a thread **yields** when it invokes a function that
+causes it to become **unready** but not terminated: that is, it
+transitions to *Waiting* or *Suspended*.
+
+**NOTE** The term "sleeping" when applied to Zephyr threads has
+historically mean a thread that is *Waiting* for a timeout event.
+
+Zephyr defines [two classes of
+threads](https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-priorities):
+
+* **cooperative** threads remain the current thread until they are made
+  unready.
+* **preemptible** threads remain the current thread until either (a) a
+  cooperative thread becomes *Ready*, or (b) a higher-priority
+  preemptible thread becomes *Ready*.
+
+For API behavior we are primarily interested in whether invoking a
+function can cause a cooperative thread to yield.
+
+### Context Terminology
+
+**TODO** fill this out
+
+This section is intended to describe the privilege and processor context
+variations in which a thread can run.  It should provide or reference
+definitions of terms like this:
+
+* Kernel [initialization
+  level](https://docs.zephyrproject.org/latest/reference/drivers/index.html#initialization-levels)
+* User-space versus kernel (system call)
+* Normal ("thread"?) versus interrupt (invoked from an Interrupt Service
+  Routine)
+
+For API behavior we are interested in whether a particular function
+**may**, **must**, or **must not** be invoked from a specific context.
+
+## Definitions
+
+### yield (thread)
+
+A thread yields when it invokes an operation that transitions it to
+*Suspended* or *Waiting*.
+
+### pre-empted (thread)
+
+A thread is pre-empted when it transitions from *Running* to *Ready*
+without having yielded.  (Only preemptible threads can be pre-empted.)
+
+### yielding (function)
+
+A function is yielding when it invokes (directly or indirectly) an
+operation that causes the current thread to yield.
+
+A function is non-yielding when it is not yielding.
+
+### thread-safe
+
+A function is thread-safe if its behavior is correct when invoked from
+multiple threads at the same time.
+
+#### Commentary
+
+Most public API will satisfy this condition; some private API may not.
```

Comment:
What condition? Thread safety? I would assume most public API is not thread safe. Interrupt safe is another matter. In general I would expect device access APIs don't automatically lock device data, since that would add unnecessary overhead to single-threaded use. It should generally be the caller's responsibility to ensure exclusive access to such data.

Reply:
I'm not sure I agree, given the existence of pre-emptible threads, and the possibility (per #20919) that cooperative threads may get booted out in the middle of whatever they're doing too.

Reply:
Clarification of the concept of *suspendable* in the latest revision supports your position; I've revised the comment.  If the sanctity of cooperative threads gets broken we'll have a problem.